### ISSUE TITLE
Provide a documented way to limit maximum file size when writing

### ISSUE THREAD

**Issue by barrywhart** [tebeka/fastavro#130]
TL;DR
* Make `Writer` class public?
* Provide a way for users of`Writer` to know each time a block is written?

We often dump very large data sets to Avro. We need a generic way to limit the output file size to something reasonable for Hadoop (say, 64-256MB). This is tricky since records can be of any size.

We found that we could use `fastavro`'s `Writer` class in the following way:

```
        writer.write(record)

        # Block count of zero indicates fastavro just reached an internal max
        # size and ran Writer.dump(). This is a good time to consider starting
        # a new file.
        if writer.block_count == 0 and output_file.tell() >= MAX_FILE_SIZE:
            output_file.close()
            writer = Writer(...)
```

This worked okay but has a couple of issues:
* The `Writer`class is not part of the public interface of the library.
* It relies on internal knowledge of the `Writer` class:
  * `block_count` and `io` fields
  * `dump()` function

Also (irony), my [recent PR](https://github.com/tebeka/fastavro/pull/114) which added `_writer.pyx` breaks the above approach because the fields are now private (due to being declared `cdef` rather than `cpdef`.

I would like to restore the above capability and am happy to contribute a PR. I'd like to hear others' thoughts on how to do this. The simplest approach would be to expose those fields to Python, but it doesn't seem like a great design. Another possible approach would update `Writer` to notify when it writes a block, perhaps by adding a new optional `on_write_block_fn` parameter to `Writer`'s constructor.

@scottbelden

**Comment by scottbelden** [comment]
Hmm... I'd be worried about exposing `Writer` for a couple of reasons.

- As you mentioned, it requires knowledge of some internals like `block_count`
- The name is basically the same as the current `writer` which could be confusing. However, the solution to that is probably to change `writer` to `write`

Having `Writer` be public doesn't seem like it would be too bad, but it might require extra thought on what parts of it should be public. For example, I don't know if `dump` should be public as it seems like `flush` should be sufficient.

One idea (not sure if I think it's good or not...) could be to have `writer` keep track of how many bytes it has written and take an option `cutoff` argument to know when it should stop.

The `writer` function would probably need to return the number of records written so that if it was given a list of records the user would know how many records still need to be written. For example:

```python
def writer(..., cutoff=None):
    output = Writer(
        fo,
        schema,
        codec,
        sync_interval,
        metadata,
        validator,
    )

    bytes_written = 0
    records_writen = 0
    for record in records:
        bytes_written += output.write(record)
        records_written += 1
        if cutoff and bytes_written > cutoff:
            break
    output.flush()
    return records_written
```

And of course `Writer.write()` would need to return the bytes written.

From a user perspective, this might look like the following:
```python
# Records as a list
records = [record1, record2, ...]

records_written = 0
for writeable_file in writeable_files:
    with open(writeable_file, 'wb') as out:
        records_written += writer(out, schema, records[records_written:], cutoff=64_000_000)
    if records_written == len(records):
        break

###

# Records as a stream
records = <some stream>

records_written = 0
for writeable_file in writeable_files:
    with open(writeable_file, 'wb') as out:
        records_written += writer(out, schema, records, cutoff=64_000_000)
    if records_written == 0:
        break
```

Overall, I'm not sure yet...

@tebeka @kkirsanov Any thoughts?

**Comment by barrywhart** [comment]
Right, my example code above relies on low-level details. I'm +1 for a better-encapsulated approach. TBH, I don't think there is any particular reason my example code calls dump(); flush() does essentially the same thing.

>One idea (not sure if I think it's good or not...) could be to have writer keep track of how many bytes it has written and take an option cutoff argument to know when it should stop.

Right, this is essentially what the example code does. Having the library do this will be useful for some users (including me), but I recognize that adding new features adds complexity to the library, affecting both users and maintainers.

So there are two ideas in play:
* Extending writer() to handle file size cutoffs
* Exposing Writer (possibly with a different name) and:
  * "file size" property
  * flush() function
  * But probably NOT the dump() function or fields

Both ideas seem reasonable.

Technically speaking, the writer() function can definitely implement the cutoff, but it feels more complex than doing so using the Writer class. I'm thinking of considerations like:
* If the file is cut off, the caller may want to call writer() again to write the remaining records to a different file.
* If passing a stream or generator to writer(), what should writer() do if an exception is thrown? This concern disappears with the class.

For these reasons, I slightly prefer the second approach, because it keeps writer() very simple and easy to understand. Lots of Python programmers won't understand things like generators, iterators and slices that well. If an application needs finer-grained control, it can use the class.

I'm trying to decide if the Writer class might be useful for other purposes. It's hard to justify exposing a new class for this one specific use case

**Comment by barrywhart** [comment]
I was thinking more about this. Writing Avro to a file should be similar to similar use cases:
* Writing to a regular file (`f = open('filename.out', 'wb'); f.write(...)`)
* Writing JSON to a file ((`f = open('filename.out', 'wb'); json.dump(...)`)
* Writing pickled objects to a file ((`f = open('filename.out', 'wb'); pickle.dump(...)`)

It is much more flexible to allow applications to push data to a file, rather than have the file "pull" data from an iterator.

The official Apache Avro library has a class for doing this. Here's an example from their [documentation](https://avro.apache.org/docs/1.8.2/gettingstartedpython.html):
```
import avro.schema
from avro.datafile import DataFileReader, DataFileWriter
from avro.io import DatumReader, DatumWriter

schema = avro.schema.parse(open("user.avsc", "rb").read())

writer = DataFileWriter(open("users.avro", "wb"), DatumWriter(), schema)
writer.append({"name": "Alyssa", "favorite_number": 256})
writer.append({"name": "Ben", "favorite_number": 7, "favorite_color": "red"})
writer.close()

reader = DataFileReader(open("users.avro", "rb"), DatumReader())
for user in reader:
    print user
reader.close()
```

Perhaps `fastavro` should support a similar API to make it easier for users to switch between the two libraries.

I also see that `avro.DataFileWriter` has a `block_count` property, so `fastavro` could expose the same property. I'm pretty sure this would satisfy my need (being able to tell when the writer is "clean" or "dirty").

If we think that exposing a class with the same name and some of the same properties as Apache's `avro` library, I'll be happy to close this PR and create a new one like that.

**Comment by tebeka** [comment]
I think the right approach for limiting is use a generator as input to `write` that will limit the number of record written. Size is tricky since it varies with compression.

Maybe we can expose a `sizeof` function that will return how many bytes a Python object will take and then the generator can be more informed about how much it has written.

**Comment by barrywhart** [comment]
That seems like a very indirect way of doing things, especially with compression having an impact as you say. Having the application check the current file size after `fastavro` writes each full block has worked well for us.

As I asked above, why does `fastavro`'s API for writing to files rely on iterators? This is very different than every common Python library I know (built-in files, JSON, pickle, Apache Avro library).

To be clear, our goal is not to enforce a hard upper limit on file size. Rather, the goal is to stop writing around a certain size. Therefore, stopping *after* writing the first block which pushed it over the limit has met our requirements perfectly.

**Comment by scottbelden** [comment]
It sounds like maybe the simplest thing to do for the moment is just to expose the `Writer` class (as the original PR did). It sounds like it can be useful as a more advanced API. I don't think the name has to change. I personally don't like the apache `avro` API and names as very java-like.

**Comment by barrywhart** [comment]
That sounds reasonable to me. I'll revert the subsequent changes, and we can continue discussion on the PR.

**Comment by scottbelden** [comment]
Resolved via https://github.com/tebeka/fastavro/pull/131

**Comment by barrywhart** [comment]
Thanks all!
